src/lobby.rs

```diff
@@ -194,11 +194,26 @@ impl SharedLobbyState {
         state.active_contributor = ActiveContributor::None;
     }
 
+    #[allow(clippy::needless_collect)]
     pub async fn clear_lobby(&self, predicate: impl Fn(&SessionInfo) -> bool + Copy + Send) {
         let mut lobby_state = self.inner.lock().await;
-        lobby_state
+        let sessions_to_remove = lobby_state
             .sessions_in_lobby
-            .retain(|_, info| !predicate(info));
+            .iter()
+            .filter_map(|(id, info)| {
+                if predicate(info) {
```

Comment:
nit: you can use `bool::then_some` here.

Example from playground:

```rust
assert_eq!(false.then_some(0), None);
assert_eq!(true.then_some(0), Some(0));
```

This is available as of Rust 1.62

bool::then has been available as of 1.50

```rust
assert_eq!(false.then(|| value), None);
```

Reply:
nifty!